parquet-column/src/main/java/org/apache/parquet/schema/Types.java

```diff
@@ -555,6 +622,741 @@ protected GroupType build(String name) {
+          "LIST is already the logical type and can't be changed");
+      return super.named(name);
+    }
+
+    public ElementBuilder<P, Q> element(PrimitiveTypeName type, Type.Repetition repetition) {
+      Preconditions
+          .checkState(elementType == null, "groupElement can be called just once on a ListBuilder");
+      if (parent != null) {
+        return new ElementBuilder<P, Q>(this, type)
+            .repetition(repetition);
+      } else {
+        return new ElementBuilder<P, Q>(this, returnType, type)
+            .repetition(repetition);
+      }
+    }
+
+    public ElementBuilder<P, Q> requiredElement(PrimitiveTypeName type) {
+      return element(type, Type.Repetition.REQUIRED);
+    }
+
+    public ElementBuilder<P, Q> optionalElement(PrimitiveTypeName type) {
+      return element(type, Type.Repetition.OPTIONAL);
+    }
+
+    public GroupElementBuilder<P, Q> groupElement(Type.Repetition repetition) {
+      Preconditions
+          .checkState(elementType == null, "groupElement can be called just once on a ListBuilder");
+      if (parent != null) {
+        return new GroupElementBuilder<P, Q>(this).repetition
+            (repetition);
+      } else {
+        return new GroupElementBuilder<P, Q>(this, returnType).repetition
+            (repetition);
+      }
+    }
+
+    public GroupElementBuilder<P, Q> requiredGroupElement() {
+      return groupElement(Type.Repetition.REQUIRED);
+    }
+
+    public GroupElementBuilder<P, Q> optionalGroupElement() {
+      return groupElement(Type.Repetition.OPTIONAL);
+    }
+
+    public MapElementBuilder<P, Q> mapElement(Type.Repetition repetition) {
+      Preconditions
+          .checkState(elementType == null, "groupElement can be called just once on a ListBuilder");
+      if (parent != null) {
+        return new MapElementBuilder<P, Q>(this).repetition(repetition);
+      } else {
+        return new MapElementBuilder<P, Q>(this, returnType).repetition(repetition);
+      }
+    }
+
+    public MapElementBuilder<P, Q> requiredMapElement() {
+      return mapElement(Type.Repetition.REQUIRED);
+    }
+
+    public MapElementBuilder<P, Q> optionalMapElement() {
+      return mapElement(Type.Repetition.OPTIONAL);
+    }
+
+    public ListElementBuilder<P, Q> listElement(Type.Repetition repetition) {
+      Preconditions
+          .checkState(elementType == null, "groupElement can be called just once on a ListBuilder");
+      if (parent != null) {
+        return new ListElementBuilder<P, Q>(this).repetition(repetition);
+      } else {
+        return new ListElementBuilder<P, Q>(this, returnType).repetition(repetition);
+      }
+    }
+
+    public ListElementBuilder<P, Q> requiredListElement() {
+      return listElement(Type.Repetition.REQUIRED);
+    }
+
+    public ListElementBuilder<P, Q> optionalListElement() {
+      return listElement(Type.Repetition.OPTIONAL);
+    }
+
+    public BaseListBuilder<P, Q> element(Type type, Type.Repetition repetition) {
```

Comment:
Done. Thanks again for catching this.

On Sun, May 17, 2015 at 2:07 PM, Ryan Blue <notifications@github.com> wrote:

> In parquet-column/src/main/java/org/apache/parquet/schema/Types.java
> <https://github.com/apache/parquet-mr/pull/148#discussion_r30474900>:
>
> > +      if (parent != null) {
> > +        return new ListElementBuilder<P, Q>(this).repetition(repetition);
> > +      } else {
> > +        return new ListElementBuilder<P, Q>(this, returnType).repetition(repetition);
> > +      }
> > +    }
> > +
> > +    public ListElementBuilder<P, Q> requiredListElement() {
> > +      return listElement(Type.Repetition.REQUIRED);
> > +    }
> > +
> > +    public ListElementBuilder<P, Q> optionalListElement() {
> > +      return listElement(Type.Repetition.OPTIONAL);
> > +    }
> > +
> > +    public BaseListBuilder<P, Q> element(Type type, Type.Repetition repetition) {
>
> It just occurred to me that repetition is ignored because the Type
> includes a repetition. Maybe we don't need requiredElement and
> optionalElement after all. Same with the key and value methods that take
> a Type.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/parquet-mr/pull/148/files#r30474900>.
>



-- 

Regards,
Ashish